hathor/websocket/streamer.py

```diff
@@ -69,17 +70,51 @@ def __init__(self,
 
         self.max_seconds_locking_event_loop = 1
 
+        # Statistics.
         self.stats_log_interval = self.STATS_LOG_INTERVAL
         self.stats_total_messages: int = 0
         self.stats_sent_addresses: int = 0
         self.stats_sent_vertices: int = 0
 
+        # Execution control.
+        self._is_running = False
         self._paused = False
         self._stop = False
 
+        # Flow control.
+        self._next_sequence_number: int = 0
+        self._last_ack: int = -1
+        self._sliding_window_size: Optional[int] = self.DEFAULT_SLIDING_WINDOW_SIZE
+
+    def set_sliding_window_size(self, size: Optional[int]) -> None:
+        """Set a new sliding window size for flow control. If size is none, disables flow control.
+        """
+        if size == self._sliding_window_size:
+            return
+        self._sliding_window_size = size
+        self.resume_if_possible()
+
+    def set_ack(self, ack: int) -> None:
+        """Set the ack value for flow control.
+
+        If the new value is bigger than the previous value, the streaming might be resumed.
+        """
+        if ack <= self._last_ack:
+            # We might receive outdated or duplicate ACKs, and we can safely ignore them.
+            return
+        if ack >= self._next_sequence_number:
+            # Should we raise an exception here?
+            return
+        self._last_ack = ack
+        self.resume_if_possible()
+
+    def resume_if_possible(self) -> None:
+        if not self.should_pause_streaming() and not self._is_running:
+            self.resumeProducing()
+
     def start(self) -> Deferred[bool]:
         """Start streaming items."""
-        self.send_message(StreamBeginMessage(id=self.stream_id))
+        self.send_message(StreamBeginMessage(id=self.stream_id, sliding_window_size=self._sliding_window_size))
```

Comment:
The client sends `window_size`, why not send `window_size` back?

Reply:
Done!